src/cascadia/WindowsTerminal/IslandWindow.cpp

```diff
@@ -115,40 +114,38 @@ void IslandWindow::Initialize()
     // stash the child interop handle so we can resize it when the main hwnd is resized
     interop->get_WindowHandle(&_interopWindowHandle);
 
-    if (!initialized)
-    {
-        _InitXamlContent();
-    }
-
+    _rootGrid = winrt::Windows::UI::Xaml::Controls::Grid();
     _source.Content(_rootGrid);
 
     // Do a quick resize to force the island to paint
     OnSize();
 }
 
-void IslandWindow::_InitXamlContent()
+double IslandWindow::GetCurrentDpiScale()
 {
     // setup a root grid that will be used to apply DPI scaling
     winrt::Windows::UI::Xaml::Media::ScaleTransform dpiScaleTransform;
     winrt::Windows::UI::Xaml::Controls::Grid dpiAdjustmentGrid;
 
     const auto dpi = GetDpiForWindow(_window);
     const double scale = double(dpi) / double(USER_DEFAULT_SCREEN_DPI);
-
-    _rootGrid = dpiAdjustmentGrid;
-    _scale = dpiScaleTransform;
-
-    _scale.ScaleX(scale);
-    _scale.ScaleY(scale);
+    return scale;
 }
 
 
 void IslandWindow::OnSize()
 {
     // update the interop window size
     SetWindowPos(_interopWindowHandle, 0, 0, 0, _currentWidth, _currentHeight, SWP_SHOWWINDOW);
-    _rootGrid.Width(_currentWidth);
-    _rootGrid.Height(_currentHeight);
+
+    if (_rootGrid)
+    {
+        const auto dpi = GetCurrentDpiScale();
+        const auto logicalWidth = (_currentWidth / dpi) + 0.5;
+        _rootGrid.Width(logicalWidth);
+        const auto logicalHeigth = (_currentHeight / dpi) + 0.5;
+        _rootGrid.Height(logicalHeigth);

```

Comment:
Any changes to OnSize here probably need to be reflected in `NonClientIslandWindow` as well -- since it resizes some other things independently.